outgoing webhooks: Make exception handling more granular and verbose.

derAnfaenger opened the pull request:
Fixes #6127.

smarx commented:
*Automated message from Dropbox CLA bot*

@derAnfaenger, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
@derAnfaenger this seems likely what we want!  But I'm not sure the message formatting is right, but I think the way to make progress on that is to add some tests for this flow.  (that function is basically untested :frowning_face:).

I think the way to do it is to write a few tests that mock `requests.request` here and e.g. raise various types of exceptions, and check the content of the message.  I suspect we'll find that the message content could be improved in many cases based on looking at those :).

derAnfaenger commented:
@timabbott will do! Btw, I reasoned that the email is being sent because of `logging.exception`? However, I couldn't find an email in the dev server output (isn't it supposed to display sent emails?), so I just assumed changing it to `logging.warning` would make it stop sending emails.

timabbott commented:
Yep, that's correct.  The flow is a bit complicated, but basically `logging.exception()` (or throwing an exception, as well) will cause the code in `zerver/logging_handlers.py` to run, which eventually calls the code in `zerver.lib.error_notify` that will send the emails.  I think probably `logging.warning` is the right level to log this at for now.  

In terms of what to log to the file, I think we want:
* The exception
* What bot and in particular URL to post to are involved

derAnfaenger commented:
@timabbott the only test I can think of adding here is one that mocks a `ConnectionError` and asserts the logging message is produced.

Regarding what to log to the file, it'd be much easier to log the involved URL once #6307 gets merged (updated it today to make it easier to review), since #6307 provides helper functions to generate the url. It'd also allow to notify the bot owner with one function call.

timabbott commented:
@derAnfaenger can you rebase this?

derAnfaenger commented:
Rebased.

timabbott commented:
In the connection error case, does recommending they look at the server logs make sense?  Also, the logs for which server?  

Also, I think it makes sense to retry ConnectionError, but not to retry 50x errors.  50x errors are unlikely to magically fix themselves in the next 5 seconds, whereas ConnectionError can be a transient network failure...  Or am I missing something here?

derAnfaenger commented:
I meant the Zulip server, will add that. Looking in the server logs would give this additional information:

> logging.warning("Trigger event %s on %s resulted in a connection error." % (event["command"], event['service_name']))

I don't see either any need for the 50x retry, will update the PR.

derAnfaenger commented:
Updated. Re-testing this also made me realize that the URL construction is broken for private messages, both in this PR and in #6307 . Going through the code, I only found the js function `by_conversation_and_time_uri` for link generation, so I'll update #6307 to correctly generate links. Once #6307 gets merged, we can simply use the function there to generate the link here.

Would it make sense to provide such a path generation function in the zulip backend code? `outgoing_webhook` seems to be the wrong place for this.

timabbott commented:
I think we could probably add this in `zerver/lib/notifications.py`; `pm_narrow_url` in that file is pretty similar.    But probably ultimately we'll want to move things from there to some new file, I think it makes sense to do that when we find `notifications.py` to be too big.

derAnfaenger commented:
Fixed the merge conflicts and updated the PR with the new notify method available. Added one tiny commit to specify `notify_bot_owner`'s mypy annotation.

timabbott commented:
Merged, after changing "Time is running..." to "Time is up!", since that's more reasonable :).

I think we'll want to invest more time into cleaning this code path up.  It doesn't at all feel like it's obviously correct, and I feel like this is a system where that should be very doable.

derAnfaenger commented:
Hm, the only thing I see that could be simplified is the various debugging endpoints: We log to the server, send a failure message to the bot user and notify the bot owner. Maybe we could merge these calls in one function? The info text seems to vary quite heavily for every case, but that could be a sign that they are not optimal yet.
